### PR TITLE
Fix ReactNativeART arc drawing on Android

### PR DESCRIPTION
**Pull request by MikeOrtiz**
**Motivation**: Arc drawing has been broken on Android for some time. @dgladkov submitted a PR, which ended up having a bug and was never merged. This PR should fix that bug as well as provide screenshots to prove it works.

**Reproducing the Bug:** dgladkov made a simple test app which helps to illustrate the bug. The repo can be found [here](https://github.com/dgladkov/RNArtArcDrawingBug). The demo app illustrates that on iOS, wedges are drawn correctly, but Android only draws full circles. [Direct Link to iOS Before](https://github.com/dgladkov/RNArtArcDrawingBug/blob/master/images/ios.png). [Direct Link to Android Before](https://github.com/dgladkov/RNArtArcDrawingBug/blob/master/images/android.png).

**Proof The Bug is Fixed:** [Here is a direct link to Android After pic.](http://i.imgur.com/9dTU2Xn.png) You can see the wedges match the iOS Before screenshot.

**What went wrong:** dgladkov's solution relied on Java's modulus, which in fact, implements modulus in a non-standard way. Modulus should always be positive. Java returns the remainder, which can be negative. `-120 % 360` will return 240 in Python, but -120 in Java. I created a modulus method which mimics Python's behavior. Math.floorMod(), introduced in Java 8, also gets the job done. See this [StackOverflow answer](http://stackoverflow.com/questions/5385024/mod-in-java-produces-negative-numbers) for more information regarding modulus.

### CODE REVIEW

**Comment by spicyj** [comment]
To be fair, lots of languages implement `%` like this.

File: ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java [hunk 1]
```diff
@@ -232,11 +246,16 @@ private Path createPath(float[] data) {
...
-          boolean clockwise = data[i++] == 0f;
+          float normalizedStart = modulus(start, 360);
+          float normalizedEnd = modulus(end, 360);
+          boolean clockwise = data[i++] == 1f;
+          float sweep = modulus(normalizedEnd - normalizedStart, 360);
+
           if (!clockwise) {
-            end = 360 - end;
+            start = end;
+            sweep = 360 - sweep;
```
Comment (spicyj) [inline ReactAndroid/src/main/java/com/facebook/react/views/art/ARTShapeShadowNode.java]: If `start === end` already and `sweep` is 0, is this correct?

**Comment by MikeOrtiz** [comment]
@spicyj Good catch. That edge case would create a full circle in counter-clockwise mode. I have a fix in place, but first a question: what is the expected result of an arc with startAngle = -360 and endAngle = 20? In my implementation, the result is a full circle, but I guess the argument could be made for a 20 degree arc. The [ART docs](https://github.com/sebmarkbage/art/blob/master/docs/ART/ART.Path.md), unfortunately, do not describe Path.arc().

**Comment by spicyj** [comment]
I think whatever SVG does.

https://github.com/sebmarkbage/art/blob/51ffce8164a555d652843241c2fdda52e186cbbd/modes/svg/path.js#L13

**Comment by MikeOrtiz** [comment]
I decided it was best to mimic the behavior of iOS. If the sweep > 360, then a full circle is always drawn. If sweep = 0, no wedge is drawn. Comparison screenshots: [iOS](http://i.imgur.com/l98XtAe.png) and [Android](http://i.imgur.com/cXGcoOf.png). The start/end angles are as follows:

**Blue**: start: 120, end: 220
**Red**: start: 120, end: 120
**Green**: start: -360, end: 20
**Orange**: start: 120, end: 20

**Comment by MikeOrtiz** [comment]
@spicyj, I realized I didn't mention you in my last post. Not sure if that makes a difference.

**Comment by MikeOrtiz** [comment]
The post isn't yet tagged as such, but I have signed the CLA.

**Comment by mkonicek** [comment]
Assigning to @spicyj who started reviewing this PR.

**Comment by tomauty** [comment]
Hooray :) This is a blocker for my project.

**Comment by spicyj** [comment]
Can you say whether this matches the SVG behavior now? I am fairly sure that is what ART intends to do for this path component.

**Comment by tomauty** [comment]
The fixed behavior seems to conform to intentions. The repo provided uses the Wedge component from react-art, in any case.

**Comment by spicyj** [comment]
Yeah, that seems fair. @facebook-github-bot shipit

**Comment by spicyj** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-0** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/476794839178941/int_phab) to review.

**Comment by spicyj** [comment]
It looks like this breaks circle rendering:

screen shot 2016-04-28 at 3 42 26 pm

```
const {Surface, Group} = require('ReactART');
const Circle = require('Circle.art');

...

  <Surface width={200} height={200}>
    <Group x={100} y={100}>
      <Circle radius={50} stroke="green" strokeWidth={3} fill="blue" />
    </Group>
  </Surface>
```

(where Circle.art comes from https://github.com/reactjs/react-art/blob/d1eda4fd8e3ee9462c32d2f611923697195fee14/src/Circle.art.js).

There should not be a line in the middle. I have to revert this commit, but would you want to send a follow-up that fixes this?

**Comment by MikeOrtiz** [comment]
Aww, yeah I'll look into it. When should I submit the fix to get it in for the upcoming release?

**Comment by spicyj** [comment]
I think we cut a release every other Sunday, including this weekend.

**Comment by MikeOrtiz** [comment]
I just ran this on my machine and it looks fine. [This is what it looks like for me.](http://i.imgur.com/lPyryqf.png) The call to close() in Circle.art is what will create the line segment to close off the final arc. If the path ends at the same point that it starts, there will be no line segment. This functionality is described in the Android Path docs [here](http://developer.android.com/reference/android/graphics/Path.html#close()). The arcs are centered around the same coordinate and the second ends at the same angle the first begins, so the starting points should be identical. I printed some logs of Circle.art to confirm that fact:

    04-29 01:54:38.488 1972-2038/com.testproject D/dsf: x: 140.0 y: 280.0 r: 140.0 start: -90.0 end: 90.0 sweep: 180.0
    04-29 01:54:38.488 1972-2038/com.testproject D/dsf: x: 140.0 y: 280.0 r: 140.0 start: 90.0 end: -90.0 sweep: 180.0

I don't understand how you got this rendering, but I can't reproduce it. 

Anyway, one simple fix would be to remove the call to close() in Circle.art as it doesn't achieve anything, anyway.

**Comment by spicyj** [comment]
Hm, I'm not sure. I was testing on a Nexus 4 simulator on 4.4.4 (API 19).

**Comment by MikeOrtiz** [comment]
@spicyj It looks like my changes weren't what broke this. The circle draws correctly up to #4e5c85 of my [repo](https://github.com/MikeOrtiz/react-native/commits/master). At some point after that, someone introduced breaking code and my change exposed that break. I will run git bisect when I have time later today to pinpoint the breaking commit.

**Comment by spicyj** [comment]
Sounds good.

**Comment by MikeOrtiz** [comment]
I couldn't successfully bisect the commits. After commit [#47a470a](https://github.com/facebook/react/pull/6338), I get the red screen of death with the message "Couldn't get the native call queue...". Restarting package manager, reinstalling, etc didn't fix the issue. The message shows for this commit and all subsequent commits. Any hint would could be going wrong?

Circle rendered fine in the commit before that though.

**Comment by spicyj** [comment]
I'm not sure, sorry. Does master work for you? (I don't really work on Android, or on React Native for that matter.)

**Pull request closed without merging**